### ISSUE TITLE
Add slice of views

### ISSUE THREAD

**Issue by mchodzikiewicz** [bugadani/embedded-layout#12]
Hi,

I am trying to use your library to draw evenly spaced views generated from an iterator. This means that I can't manually add views by `.add_view`, because number of views depends on axis config, and I cannot do this inside `.fold` because the type is changing with each iteration.
Ideally I'd like to do something similar:
```rust
    (0..10).into_iter()
        .fold(
            LinearLayout::horizontal(),
            |layout, _line|
                layout.add_view(
                    Line::new(
                        Point::new(0, 0),
                        Point::new(0, 10))
                        .into_styled(PrimitiveStyle::with_stroke(RgbColor::GREEN, 1))))
        .arrange()
        .align_to(&display.display_area(),horizontal::Center,vertical::Center)
        .draw(&display);
```
but I know that this is not possible to overcome this.

Perhaps passing a slice of views would be possible? Or separate module for aligning repeated views?

My ultimate goal is to refactor/rewrite following snippet, so the axis is drawn inside the provided rectangle;
https://gitlab.com/mchodzikiewicz/embedded-plots/-/blob/master/src/axis.rs#L171

**Comment by bugadani** [comment]
Hi! I have to admit the current implementation is a dead-end, but I haven't put any time into coming up a solution yet. I'd definitely like to replace the current type-level chaining with something more flexible!

I'll try to come up with something soon(-ish).

Oh, and embedded-plots is awesome, congrats!

**Comment by mchodzikiewicz** [comment]
Thank you! Long way before it'll be a production grade one :)

Current solution is very useful for more high level, predetermined layouts but this is kind of a generic problem that solved would make embedded-layout suitable for reusing in other generic crates.

For me, I think I have solved this for now manually, leaving calculation of ticks done by hand and enclosing text by TextBoxes from `embedded-text`, but I'll also try to think about it myself. Will get back here if I find something useful.

**Comment by bugadani** [comment]
Interesting... I have an idea of what to do. The original solution's inflexibility comes from the existence of `Layout::add_view`. If that could be replaced by a `Layout::from_view_group`-like constructor, anything could be wrapped into a layout.

I'm considering something like this (fair warning: [🧠🌧] completely untested and the idea may be useless, will need to play around):
```rust
trait ViewGroup: View + Index<u32, Output = dyn View> + IndexMut<u32> {
    fn count(&self) -> u32;
}

struct LinearLayout<VG: ViewGroup>(ViewGroup);

impl<VG: ViewGroup> LinearLayout {
    pub fn new(vg: VG) -> Self {
        Self(vg)
    }
}

impl AsRef<ViewGroup> for LinearLayout {...} // to not lose access to members
...

impl<T: View> ViewGroup for [T; <N>] { ... }
impl<A: View, B: View> ViewGroup for (A, B) {
    // tuples are interesting, I have no idea rn if this would work for tuple trees
    // but they seem somewhat composable
}
```

Even with these ideas, what's in the OP would look different. But in your case, as long as you know how many ticks you have, you can build an array first and pass that to a LinearLayout. Not sure if this is something close enough to what you need.

**Comment by mchodzikiewicz** [comment]
Example I have provided was just to show the limitation, any possible clean way to achieve this would be sufficient.

I think it's a good starting point to explore.
I'd rather use [Len](https://docs.rs/len-trait/0.2.2/len_trait/trait.Len.html) trait instead of `fn count() -> u32`

I think current typechaining approach is good enough if from API perspective  you could do either `add_view` or `add_view_group` taking a `ViewGroup` which is (at least - perhaps other constrains apply) `IntoIterator where Item=View` and `Len` and embedding such collection to this typechain. Do you think this is possible?

**Comment by bugadani** [comment]
To be honest, `Len` doesn't look useful here - looks like it's not implemented for `heapless::Vec` for example. That trait seems more trouble than worth. But I can be wrong here.

I'm not sold on iterators, either. I was thinking about IntoIterator, but I think random access will be more flexible. Certainly less moving pieces to implement, at least it seems that way now. Keep in mind that LinearLayout is a rather special case, where it only needs to walk through views in order twice, but other algorithms can perhaps be arbitrarily complex.

**Comment by mchodzikiewicz** [comment]
Well.. you might be right, not sure which trait would be the best to indicate what is required from the View, I guess [SliceIndex](https://doc.rust-lang.org/core/slice/trait.SliceIndex.html) might be ok?

Am I correct it is implemented because `heaples::Vec` implements [GenericArray](https://docs.rs/generic-array/0.14.4/generic_array/iter/struct.GenericArrayIter.html#method.as_slice)? It seems to be most primitive source of this property I could find.

Also - if possible, I'd avoid over-restricting View type and create separate traits only with absolutely necessary properties - for example, you would force all styled objects to be allocated on the stack (which for linear layout is not needed) prior to drawing your layout, where if using iterator, you can write `Iterator` that generates objects to draw on the fly. 

Real life example: bitmap read from SPI external flash

Sorry if I miss the point, I am still a bit of a Rust newcomer.

**Comment by bugadani** [comment]
@mchodzikiewicz I've implemented some of my ideas [here](https://github.com/bugadani/embedded-layout/pull/13). Trait objects are a pain, but I'm pleased to say it's at least possible to use other collections as viewgroups as well.

You might be right about `heapless::Vec`. I'm still not sold on `Len`, but I haven't put much thought into the question.

> you would force all styled objects to be allocated on the stack

I don't think I am (implementing Index doesn't mean all views must be in memory, not at all!), not any more than using an Iterator would. Also note that if you have memory constraints, you should probably only read the bitmap from flash in your drawable implementation - but I haven't really worked with any bitmaps in Rust. I don't think it matters for that how ViewGroup is defined :)

I *guess* we could go with `IntoIterator` but I don't see how it would be a win - even LinearLayout needs to walk through the views twice. Also, you can implement `ViewGroup` for a struct that constructs your views in `Index`. You'd need to store that view somewhere, but I think you'd have to do that with an Iterator as well (as you can't return references to temporary values).

If you want, you can check the PR now, but things will change probably still.

**Comment by mchodzikiewicz** [comment]
Yeah... I know - at least it is still more pleasant than C++ meta programming since lifetime checks are in place :slightly_smiling_face: 

You are right about indexing - I guess there still might be some cases where random access is not possible, but I cannot figure out any practical example. I think it will still be easier to refactor new approach to accept iterators so I'd call it a day until particular use case will appear.

I have quickly checked PR's examples - if the API will stay similar (generic `ViewGroup` passed by constructor) it looks very promising

**Comment by bugadani** [comment]
@mchodzikiewicz So techincally this doesn't concern your issue, but maybe you find it interesting:

I've added a derive macro to generate ViewGroup implementations for "normal" structs. For some reason I had to replace `Index + IndexMut` with methods on the `ViewGroup` trait, but I don't really know why. At any rate, there's a [new example](https://github.com/bugadani/embedded-layout/blob/ecb89d6dd49e1975dc6b44c2fa26c209242ab94c/examples/linear_layout_custom_view_group.rs) showing how to use it. I think it's comfortable enough, although I'm not sure the new macro should be responsible for everything I'm asking it to do.

The implementation could be nicer, but I was just aiming for it to work.

Late edit: I've also pushed a commit with a wrapper struct for slices. Unfortunately it's not possible to directly implement `ViewGroup for &'a [T]` (due to possible future conflicting implementations), but a wrapper struct works well enough, I think.
